### ISSUE TITLE
Same warning printed twice

### ISSUE THREAD

**Issue by okkdev** [gleam-lang/gleam#3899]
My `list.concat` deprecation warning is being printed twice for the same line/occurrence.
```
warning: Deprecated value used
    ┌─ /Users/js/Desktop/scream/src/scream.gleam:101:12
    │
101 │     |> list.concat
    │            ^^^^^^^ This value has been deprecated

It was deprecated with this message: Use `list.flatten` instead.

warning: Deprecated value used
    ┌─ /Users/js/Desktop/scream/src/scream.gleam:101:12
    │
101 │     |> list.concat
    │            ^^^^^^^ This value has been deprecated

It was deprecated with this message: Use `list.flatten` instead.
```

## Reproduction
```sh
git clone git@github.com:okkdev/scream.git
cd scream/
git checkout dfde544ea477f2e2c27732596c1ab1397612cf41
gleam build
```

## Versions
- Gleam 1.6.1
- macOS 14.6.1
- javascript and erlang target

## Log
[log.txt](https://github.com/user-attachments/files/17921623/log.txt)

**Comment by GearsDatapacks** [comment]
Looking into this more, it's very strange.

This code prints the warning twice:
```gleam
import gleam/list

pub fn main() {
  use <- fn(f) { f() }

  list.concat([])
}
```
However, this doesn't:
```gleam
import gleam/list

fn usable(f) { f() }

pub fn main() {
  use <- usable

  list.concat([])
}
```

It only seems to occur if the deprecated value is used inside a `use` callback, where the right-hand side of the `use` expression itself is a function literal.

It also prints more times if in several nested `use` callbacks, doubling each time:
```gleam
import gleam/list

pub fn main() {
  use <- fn(f) { f() }
  use <- fn(f) { f() }
  use <- fn(f) { f() }

  list.concat([])
  //   ^ This deprecation warning prints 16 times!
}
```

It does not seem to matter which deprecated value is used.

**Comment by lpil** [comment]
>  ^ This deprecation warning prints 16 times!

Oops, ahahaha

I presume we're checking for deprecations in the body once before and once after use is expanded?

**Comment by GearsDatapacks** [comment]
Yes that's what I thought too. Could be related to the recent changes to the ast representation of use, when the desugar code action was introduced. Or this could have always been the case, just never noticed before. I'll test on 1.6.2 and then bisect it later.

**Comment by GearsDatapacks** [comment]
Ok, so it looks like it was first introduced in 1.5 in d022ff05f2ac5126292e69eb46d725b2eadaf72a. My guess is that `infer_fn_with_call_context` does a second pass over the code and produces duplicate warnings. I'll try and fix this when I get time

**Comment by lpil** [comment]
https://github.com/gleam-lang/gleam/commit/d022ff05f2ac5126292e69eb46d725b2eadaf72a#diff-892d21f7b2f39e0c5d4ef6a85f69d96b223f6cd4309106ac900cf9c391abb183R2911

Looks like it's doing some wasteful cloning there too? It clones the call arg but discards the original?

**Comment by GearsDatapacks** [comment]
Ok, so looking into this issue more, it's a bit tricky to figure out a way to get this to work. Basically:
- To infer the arguments, we first infer the left side of the call
- To be able to determine the types of unannotated arguments in an anonymous function, we first infer the arguments

I'll have a play around with it a bit more to try and figure out a solution without breaking any existing behaviour

**Comment by jtdowney** [comment]
I have a branch where I took a pass at fixing this, but wanted to share my approach before submitting a PR. In `infer_fn_with_call_context`, when the call argument is the implicit callback inserted by use desugaring (and is a `Fn`), I skipped the speculative inference by building a `fn(_, ...) -> _ `from fresh unbound vars matching the callback's arity instead, and let `do_infer_call_with_known_fun` do the real inference in the second pass. The speculative pass would be preserved for non-fn-literal args since #2504 still needs it to type unannotated params from the call context.

I did not look into cleaning up the wasteful clones, but I can do that in a different pass.

**Comment by giacomocavalieri** [comment]
Mhm it's a bit hard for me to understand the fix, could you maybe make a draft pull request so we can have a look at it?

**Comment by lpil** [comment]
Perhaps we can pass on this for now as we are deprecating the ambiguous pipe?

**Comment by GearsDatapacks** [comment]
I don't see how those two features are related. This bug occurs with `use` and regular calls as well as pipes.

**Comment by lpil** [comment]
Oh, sorry! I thought this was due to analysing pipe expressions twice. I didn't ready the whole thread recently.

**Comment by jtdowney** [comment]
This is about typing the unannotated params of a called closure from the call's arguments. The speculative pass over the args emits the warning once, because the arg is itself a closure whose body contains the warning, and then the real pass over the same args emits it a second time.

I initially thought I had a clever way to get around it, but it turned out not to be a complete solution. I opened #5944 with my proposed fix, which stashes and restores the state after inferring the argument types.
